Exception in agent logs if no tools installed on agent

```text
[2024-12-11 12:08:03,427]   WARN - ildServer.util.EventDispatcher - Error in event handler: Error calling method AgentLifeCycleListener.beforeAgentConfigurationLoaded for listener jetbrains.buildServer.rust.CargoToolProvider: java.util.NoSuchElementException
[2024-12-11 12:08:03,427]  DEBUG - ildServer.util.EventDispatcher - Error in event handler
Error calling method AgentLifeCycleListener.beforeAgentConfigurationLoaded for listener jetbrains.buildServer.rust.CargoToolProvider: java.util.NoSuchElementException
	at jetbrains.buildServer.util.EventDispatcher.invokeListeners(EventDispatcher.java:175)
	at jetbrains.buildServer.util.EventDispatcher.lambda$dispatch$0(EventDispatcher.java:134)
	at jetbrains.buildServer.util.NamedThreadFactory.executeWithNewThreadName(NamedThreadFactory.java:76)
	at jetbrains.buildServer.util.EventDispatcher.dispatch(EventDispatcher.java:134)
	at jetbrains.buildServer.util.EventDispatcher$2.invoke(EventDispatcher.java:82)
	at com.sun.proxy.$Proxy28.beforeAgentConfigurationLoaded(Unknown Source)
	at jetbrains.buildServer.agent.impl.BuildAgentImpl.init(BuildAgentImpl.java:896)
	at jetbrains.buildServer.agent.AgentMain2.startAgent(AgentMain2.java:222)
	at jetbrains.buildServer.agent.AgentMain2.main2(AgentMain2.java:98)
	at jetbrains.buildServer.agent.AgentMain2.main(AgentMain2.java:42)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at jetbrains.buildServer.agent.AgentMain.main(AgentMain.java:44)
Caused by: java.util.NoSuchElementException
	at jetbrains.buildServer.rust.AbstractToolProvider.findToolPath(AbstractToolProvider.kt:148)
	at jetbrains.buildServer.rust.AbstractToolProvider.beforeAgentConfigurationLoaded(AbstractToolProvider.kt:36)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at jetbrains.buildServer.util.EventDispatcher.invokeListeners(EventDispatcher.java:157)
	... 14 more
```

--- a/plugin-rust-agent/src/main/kotlin/jetbrains/buildServer/rust/AbstractToolProvider.kt
+++ b/plugin-rust-agent/src/main/kotlin/jetbrains/buildServer/rust/AbstractToolProvider.kt
@@ -16,8 +16,10 @@ import jetbrains.buildServer.SimpleCommandLineProcessRunner
 import jetbrains.buildServer.agent.*
 import jetbrains.buildServer.util.EventDispatcher
 import jetbrains.buildServer.util.StringUtil
-
 import java.io.File
+import java.nio.file.Files
+import java.nio.file.Path
+import kotlin.io.path.absolutePathString
 
 /**
  * Determines tool location.
@@ -31,9 +33,12 @@ abstract class AbstractToolProvider(
 ): AgentLifeCycleAdapter(), ToolProvider {
     private val LOG = Logger.getInstance(AbstractToolProvider::class.java.name)
     private val VERSION_PATTERN = Regex("^$configName[\\s-]([^\\s]+)", RegexOption.IGNORE_CASE)
-    private val PATH_PATTERN = Regex("^.*$configName(\\.(exe))?$", RegexOption.IGNORE_CASE)
+    private val FILE_NAME_PATH_FILTER: (Path) -> Boolean = { path: Path ->
+        path.fileName.toString().lowercase().removeSuffix(".exe") == configExecutableName
+    }
 
     init {
+        require(configExecutableName.lowercase() == configExecutableName)
         toolsRegistry.registerToolProvider(this)
         events.addListener(this)
     }
@@ -85,18 +90,29 @@ abstract class AbstractToolProvider(
             System.getenv(CargoConstants.ENV_CARGO_HOME)?.let {
                 this.add(it + File.separatorChar + "bin")
             }
-            this.addAll(StringUtil.splitHonorQuotes(System.getenv("PATH"), File.pathSeparatorChar))
-            add(System.getProperty("user.home") + File.separatorChar + ".cargo" + File.separatorChar + "bin")
+            System.getenv("PATH")?.let {
+                this.addAll(StringUtil.splitHonorQuotes(it, File.pathSeparatorChar))
+            }
+            System.getProperty("user.home")?.let {
+                this.add(it + File.separatorChar + ".cargo" + File.separatorChar + "bin")
+            }
         }
 
-        return paths.mapNotNull { File(it).listFiles() }
-                .flatMap { it.map { it.absolutePath } }
-                .filter { PATH_PATTERN.matches(it) }
-                .mapNotNull {
+        return paths.map { Path.of(it) }
+            .filter { Files.isDirectory(it) }
+            .flatMap { dir ->
+                try {
+                    Files.newDirectoryStream(dir, FILE_NAME_PATH_FILTER).use { stream -> stream.toList() }
+                } catch (_: Throwable) {
+                    emptyList<Path>()
+                }
+            }
+            .map { it.absolutePathString() }
+            .mapNotNull {
                     try {
                         val commandLine = getVersionCommandLine(it)
                         val result = SimpleCommandLineProcessRunner.runCommand(commandLine, byteArrayOf())
-                        val version = VERSION_PATTERN.find(result.stdout)?.destructured?.component1() ?: result.stdout
+                        val version = VERSION_PATTERN.find(result.stdout)?.groups?.get(1)?.value ?: result.stdout
                         it to Version.valueOf(version)
                     } catch (e: Throwable) {
                         LOG.warnAndDebugDetails("Failed to parse $configName version: ${e.message}", e)

